src/blas/backends/cublas/cublas_task.hpp

```diff
@@ -0,0 +1,33 @@
+#ifndef _MKL_BLAS_CUBLAS_TASK_HPP_
+#define _MKL_BLAS_CUBLAS_TASK_HPP_
+#include <cublas_v2.h>
+#include <cuda.h>
+#include <complex>
+#include <CL/sycl.hpp>
+#include "oneapi/mkl/types.hpp"
+#include "cublas_scope_handle.hpp"
+#include <CL/sycl/detail/pi.hpp>
+
+namespace oneapi {
+namespace mkl {
+namespace blas {
+namespace cublas {
+
+template <typename H, typename F>
+static inline auto host_task_internal(H &cgh, cl::sycl::queue queue, F f) -> decltype(cgh.interop_task(f)) {
```

Comment:
the decltype here looks fishy.. first of all, you don't return anything, did you mean to write `return cgh.interop_task([f,queue]..)`?
secondly, if making `decltype(cgh.interop_task(f))` shouldn't work at all, as `interop_task` would not be able to instantiate `f`, if it doesn't take an `interop_handler` as parameter.. right?

so either replace the auto and decltype.. with void (as you are discarding the result in `onemkl_cublas_host_task` anyways or you have to use something like `decltype(cgh.interop_task([](cl::sycl::interop_handler ih))` (you could also change the parameter type to auto, I think..)

Reply:
Your totally right, this part of the code is quite confusing. I followed the pattern that has been used for the CPU functions see: https://github.com/oneapi-src/oneMKL/blob/e8e3dabf9fbda0556b8075c76b657336f88440f0/src/blas/backends/mklcpu/mklcpu_common.hpp#L42-L56 

Probably would be nicer to do something like:
```
template <typname H, typename F, deceltype(cgh.interop_task(f))
.... void host_task_internal...
```
I intend to add the functionality for hipSYCL like this:
https://github.com/sbalint98/oneMKL/blob/test-int3/src/blas/backends/cublas/cublas_task.hpp

Reply:
Your totally right, this part of the code is quite confusing. I followed the pattern that has been used for the CPU functions see: https://github.com/oneapi-src/oneMKL/blob/e8e3dabf9fbda0556b8075c76b657336f88440f0/src/blas/backends/mklcpu/mklcpu_common.hpp#L42-L56 

Probably would be nicer to do something like:
```
template <typname H, typename F, deceltype(cgh.interop_task(f))
.... void host_task_internal...
```
I intend to add the functionality for hipSYCL like this:
https://github.com/sbalint98/oneMKL/blob/test-int3/src/blas/backends/cublas/cublas_task.hpp

Reply:
What does it even return? I think none of the `handler` member functions that are there to be called inside `queue::submit` have a non-void return type in SYCL. I imagine this is also the case for `interop_task` - so why can the return type not simply be `void`?

Reply:
It's basically sfinae on whether the function exists or not...
For CPU they already use something similar in oneMKL, but there the interface of f fits the requirements of the task function, which is not the case here(taking the scoped thing instead of a handle(r), but you can't wrap it in a lambda either, as it's in an unevaluated context. )
It probably works as log the checked functions don't use sfinae to check the signature of the parameter..